### PR TITLE
feat(elements): Add initial sign-up

### PR DESCRIPTION
**Pull request by tmilewski**
## Description

@BRKalow Just what we reviewed live on Friday.

SDK-1136 - Sign Up Base Machine
SDK-1147 - Sign Up Start
SDK-1148 - Sign Up Social Buttons
SDK-1150 - Sign Up Form
SDK-1151 - Sign Up Verification: Email Link

## Checklist

- [x] `npm test` runs as expected.
- [x] `npm run build` runs as expected.
- [ ] (If applicable) [JSDoc comments](https://jsdoc.app/about-getting-started.html) have been added or updated for any package exports
- [ ] (If applicable) [Documentation](https://github.com/clerk/clerk-docs) has been updated

## Type of change

- [ ] 🐛 Bug fix
- [x] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

## Packages affected

- [ ] `@clerk/backend`
- [ ] `@clerk/chrome-extension`
- [ ] `@clerk/clerk-js`
- [ ] `@clerk/clerk-expo`
- [ ] `@clerk/fastify`
- [ ] `gatsby-plugin-clerk`
- [ ] `@clerk/localizations`
- [ ] `@clerk/nextjs`
- [ ] `@clerk/clerk-react`
- [ ] `@clerk/remix`
- [ ] `@clerk/clerk-sdk-node`
- [ ] `@clerk/shared`
- [ ] `@clerk/themes`
- [ ] `@clerk/types`
- [ ] `build/tooling/chore`

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -1,5 +1,12 @@
 {
-  "search.exclude": {},
+  "search.exclude": {
```
Comment (tmilewski) [inline .vscode/settings.json:2]: Just helpful default search exclusions, toggled by:
Screenshot 2024-01-16 at 1 32 14 PM

File: packages/elements/src/internals/constants/clerkjs.ts
Comment (BRKalow) [inline packages/elements/src/internals/constants/clerkjs.ts:1]: nit: I would not have expected the contents of this file to live under `/constants` given it contains a bunch of types and some type guards 💭

**Review by BRKalow — APPROVED** [review]
lgtm, left some minor comments

File: packages/elements/src/internals/machines/sign-up/sign-up.context.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+// ================= SELECTORS ================= //
+
+/**
+ * Selects the clerk environment
+ */
+const clerkEnvironmentSelector = (state: SnapshotState) => state.context.clerk.__unstable__environment;
+
+// ================= HOOKS ================= //
+
+export function useStrategy(name: SignInStrategyName) {
```
[the remark this answers is not in the document]
Comment (tmilewski) [reply packages/elements/src/internals/machines/sign-up/sign-up.context.ts:50]: This is being removed in the follow-up PR.

File: packages/elements/src/index.ts [hunk 3]
```diff
@@ -17,6 +17,19 @@ export {
   SignInStrategy,
 } from '~/react/sign-in';
 
+/** Sign Up Components */
+export {
+  SignUp,
+  SignUpStart,
+  // SignUpFactorOne,
+  // SignUpFactorTwo,
```
[the remark this answers is not in the document]
Comment (tmilewski) [reply packages/elements/src/index.ts:25]: This is being removed in the follow-up PR.

File: packages/elements/src/react/sign-up/index.tsx [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+  useStrategy,
+} from '~/internals/machines/sign-up/sign-up.context';
+import { Form } from '~/react/common/form';
+import { Router, useClerkRouter, useNextRouter } from '~/react/router';
+import { createBrowserInspectorReactHook } from '~/react/utils/xstate';
+import type { ThirdPartyStrategy } from '~/utils/third-party-strategies';
+
+const { useBrowserInspector } = createBrowserInspectorReactHook();
+
+// ================= SignInFlowProvider ================= //
```
[the remark this answers is not in the document]
Comment (tmilewski) [reply packages/elements/src/react/sign-up/index.tsx:23]: I'm cool with changing the structure or doing away with them altogether. 

If it's the latter, I'd love to consider how we're organizing the code as I generally just use these comments as a visual indicator of where one set of related code/types/etc. starts and ends.

File: packages/elements/src/internals/machines/sign-up/sign-up.machine.ts [hunk 5]
```diff
@@ -0,0 +1,600 @@
...
+                onDone: {
+                  actions: 'assignResource',
+                  target: 'Attempting',
+                },
+              },
+            },
+            Attempting: {
+              on: {
+                'EMAIL_LINK.VERIFIED': {
+                  // event.output.verifications.emailAddress.verifiedFromTheSameClient (TODO: MODAL?)
```
[the remark this answers is not in the document]
Comment (tmilewski) [reply packages/elements/src/internals/machines/sign-up/sign-up.machine.ts:451]: Possibly, yes.

**Pull request merged**